rust/src/crypto.rs

```diff
@@ -0,0 +1,314 @@
+use derive_more::{AsMut, AsRef, From};
+use sodiumoxide::crypto::{box_, sealedbox, sign};
+
+pub trait ByteObject: Sized {
+    /// Create a new object with all the bytes initialized to `0`.
+    fn zeroed() -> Self;
+
+    /// Get the object byte representation
+    fn as_slice(&self) -> &[u8];
+
+    /// Create a object from the given buffer. This function will fail
+    /// and return `None` if the length of the byte-slice isn't equal to
+    /// the length of the object.
+    fn from_slice(bytes: &[u8]) -> Option<Self>;
+
+    /// Create a object from the given buffer.
+    ///
+    /// # Panic
+    ///
+    /// This function will panic if the length of the byte-slice isn't
+    /// equal to the length of the object.
+    fn from_slice_unchecked(bytes: &[u8]) -> Self {
+        Self::from_slice(bytes).unwrap()
+    }
+}
+
+/// Generate a new random key pair
+pub fn generate_encrypt_key_pair() -> (PublicEncryptKey, SecretEncryptKey) {
+    let (pk, sk) = box_::gen_keypair();
+    (PublicEncryptKey(pk), SecretEncryptKey(sk))
+}
+
+/// Deterministically derive a new key pair from the given seed
+pub fn derive_key_pair_from_seed(seed: &KeySeed) -> (PublicEncryptKey, SecretEncryptKey) {
```

Comment:
an optional suggestion: this could become a function of `EncryptKeySeed` like the following
```
impl EncryptKeySeed {
    pub fn derive_encrypt_key_pair(&self) -> (PublicEncryptKey, SecretEncryptKey) {
        let (pk, sk) = box_::keypair_from_seed(self.as_ref());
        (PublicEncryptKey(pk), SecretEncryptKey(sk))
    }
}
```
and similarly for signing seeds

Reply:
good idea